### ISSUE TITLE
Gradient not accumulated across nodes

### ISSUE THREAD

**Issue by yahyunee** [deepspeedai/DeepSpeed#7419]
Hello, I'm trying to make a deepspeed version of a code that worked without  deepspeed and see if the results can be replicated in deepspeed version. However, it seems our code is not working properly and hence wanted to ask for guidance. 

TLDR:

1. **Cannot access gradient from model_engine** (only from the model itself which model_engine wraps)
2. **Gradient does not seem to be aggregated across gpus after `.backward`** (i.e. when we run with 4 gpus with batch size of 32 each, the batch size of the input is 32 while the batch size of the gradient is also 32)
3. The validation (eval mode) loss for the same data is the same for all ranks except rank 0.

We would greatly appreciate if anyone could give us even a slight hint into debugging this.

Below is the a detailed explanation.

### Background 

Deepspeed version and no deepspeed version code exists separately.
- deepspeed version code: `pretrain_main_deepspeed.py`, `pretrain_trainer_deepspeed.py`
- no deepspeed version code : `pretrain_main.py`, `pretrain_trainer.py`

Experimental conditions are
- IDENTICAL :  global batch size, lr, wd, similar optimizer(AdamW , FusedAdam) 
-  DIFFERENT : number of gpus
=> We expect them to give same training curves.

I found out that 1node 4GPU version(deepspeed) seem to work fine based on validation loss, following the loss curve of 1node 4gpu(torch’s Data Parallel). (Orange : deepspeed, Blue: no deepspeed)

<img width="800" height="477" alt="Image" src="https://github.com/user-attachments/assets/72d9db5e-4f63-496b-b505-8c662327219a" />
 
However, in 16 node condition, loss doesn't converge to expected level. I've tried variants of below.

- (a) same effective batch size : batch_size 2 (effective batch size = 2 * 16Nodes * 4GPUs = 128)
- (b) batch_size(same batch_size per GPU : 128 / batch_size per GPU : 32) X lr (optimal lr in 1node condition * **N** / ***sqrt(N)**)
Below is result of (b).

<img width="1029" height="532" alt="Image" src="https://github.com/user-attachments/assets/8424e636-5956-4e8a-ad86-b3cd1d835863" />

### Implementation details
I've attached the script I've used. 
Note) `train_micro_batch_size_per_gpu` is not used in the actual code, argument `batch_size` is the actual batch size fed into dataloader.
```
    data_loader = DataLoader(
        train_dataset, batch_size=params.batch_size, # This is per-GPU batch size
        sampler=train_sampler,
        num_workers=16, shuffle=(train_sampler is None), # Shuffle is True if not using DDP sampler
        persistent_workers=True,
        prefetch_factor=3, pin_memory=True, drop_last=True, collate_fn=collate_fn_for_data_info
    )
```

[run_deepspeed_CBraMod_pretraining_lucy.sh.zip](https://github.com/user-attachments/files/21115093/run_deepspeed_CBraMod_pretraining_lucy.sh.zip)
[pretrain_trainer_deepspeed.py.zip](https://github.com/user-attachments/files/21115665/pretrain_trainer_deepspeed.py.zip)

### Problem & Questions
So I'm guessing there's fundamental problem with my deepspeed code.
I've printed out below to see if gradient is accumulated across ranks.
There were three problems
1. Cannot inspect gradient from `model_engine`
2. Gradient shape not as expected
3. Validation loss with fake data gave different loss only for rank 0.
 
#### Questions
1. How to inspect gradient across ranks? 
I've tried two versions as below. `model` printed out gradients while `model_engine` didn't. (2node 4GPU)
 
       def debug_gradients(self, batch_idx, log_every=10):
            """Quick gradient debugging"""
            if batch_idx % log_every != 0:
                return

        rank = deepspeed.comm.get_rank() if self.params.deepspeed else 8888
        print(f"\n[Rank {rank}] Gradient Check - Batch {batch_idx}")
        
        print("✅MODEL VERSION")
        for name, param in self.model.named_parameters():
            grad = deepspeed.utils.safe_get_full_grad(param)
            if grad is not None:
                print("Gradient Shape: ", name, grad.shape)
            else:
                print("Gradient is None for parameter:", name)
                
        print("✅MODEL ENGINE VERSION")       
        for name, param in self.model_engine.named_parameters():
            if param.grad is not None:
                grad_norm = param.grad.norm().item()
                print(f"  {name}: shape={param.grad.shape}, norm={grad_norm:.4f}")
            else:
                print(f"  {name}: No gradient")`

```

x.shape:  torch.Size([32, 19, 30, 500])

✅MODEL VERSION
Gradient is None for parameter: mask_encoding
Gradient Shape:  embedding.0.proj_in.0.weight torch.Size([32, 1, 1, 63])
Gradient Shape:  embedding.0.proj_in.0.bias torch.Size([32])
Gradient Shape:  embedding.0.proj_in.1.weight torch.Size([32])
Gradient Shape:  embedding.0.proj_in.1.bias torch.Size([32])
Gradient Shape:  embedding.0.proj_in.3.weight torch.Size([32, 32, 1, 3])

✅MODEL ENGINE VERSION
  module.mask_encoding: No gradient
  module.embedding.0.proj_in.0.weight: No gradient
  module.embedding.0.proj_in.0.bias: No gradient
  module.embedding.0.proj_in.1.weight: No gradient
  module.embedding.0.proj_in.1.bias: No gradient
```
2.  Also, input shape, gradient shape weren't as I expected.
(1node 4GPU)
Since batch_size is 32, I think gradient shape should be (128, -,-,-) since gradient should be shared across model_engine. How should I target this?

3.  I've checked validation loss with fake data (same shape with input x, `torch.ones` and `torch.zeros`) and found validation loss differ only in rank 0. (2node 4GPU)
<Validation code>
`def validate(self, epoch, normalize_factor=100.0):
        self.model_engine.eval()
        valid_losses_rank = [] # Losses on this rank

        # tqdm only on rank 0
        iterable_valid_loader = self.valid_data_loader
        if self.is_rank_0:
            iterable_valid_loader = tqdm(self.valid_data_loader, desc=f"Validation Epoch {epoch}", mininterval=10)

        with torch.no_grad():
            for batch_idx, (x, data_info_list) in enumerate(iterable_valid_loader):
                x = x.to(self.device, dtype=torch.float32) / 100.0
                print("validation x.shape: ", x.shape)
                
                ##!
                fake_data = torch.ones_like(x, device=self.device, dtype=torch.float32)/2 # Create (same) fake data across ranks
                fake_loss = self.SSL.compute_loss(fake_data, data_info_list=data_info_list)
                print(f"Fake loss for validation for rank {deepspeed.comm.get_rank()}: {fake_loss.item()}") # Print fake loss for debugging
                ##!`

```
Fake loss for validation for rank 0: 0.8215165138244629
Fake loss for validation for rank 4: 0.8218275308609009
Fake loss for validation for rank 5: 0.8218275308609009
Fake loss for validation for rank 7: 0.8218275308609009
Fake loss for validation for rank 6: 0.8218275308609009
Fake loss for validation for rank 2: 0.8218275308609009
Fake loss for validation for rank 1: 0.8218275308609009
Fake loss for validation for rank 3: 0.8218275308609009
```

**Comment by sfc-gh-truwase** [comment]
@yahyunee thanks for reporting these issues. Below are my initial thoughts and requests. 

1. It is quite strange that DeepSpeed works for single node but breaks with multi-node. It would be useful to get more details about this:
   a. What is the deepspeed config used in these experiments?
   b. Did you consider an intermediate scale, such as 2 nodes?
2. As documented [here](https://deepspeed.readthedocs.io/en/latest/zero3.html#debugging), the API for examining params and gradients works on model not model_engine. Is there any reason to use model_engine?
3. Regarding gradient shape, my understanding is that gradients have the same shape as the param and unrelated to input or batch shape. Can you please confirm that you are seeing a difference between deepspeed and ddp on this?
4. In general, it is most useful to provide code, data, and scripts to allow full side-by-side repro of with and without deepspeed. It can also be useful to share these artifacts as github repo or gist. It was challenging for me to analyze the codes that you shared due to the unrelated aspects. Being able to run the code would help focus on the pertinent parts.

**Comment by yahyunee** [comment]
Thank you for your reply and so sorry for my late reply.

Actually the issue was resolved (== the loss curve was identical for single node and multi-node setting) when I set global batch size identically. It should have been micro batch size 2 for 16 node condition not micro batchsize 8. 

The issue was resolved but I'll provide some details since it might be helpful for ones who are encountering similar situation.

1. I've also tried 2nodes this time and it showed identical loss curve! Config was below.
`{
    "train_micro_batch_size_per_gpu": 2,
    "steps_per_print": 2000,
    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": $lr,
            "betas": [0.8, 0.999],
            "eps": 1e-8,
            "weight_decay": $wd
        }
    },
    "gradient_clipping": 1.0,
    "zero_optimization": {
        "stage": 1
    },
    "fp16": {
        "enabled": false
    },
    "flops_profiler": {
        "enabled": true,
        "profile_step": 1,
        "module_depth": -1,
        "top_modules": 1,
        "detailed": true,
        "output_file": null
    }
}`
2. There was no reason to use model_engine but it was due to my misunderstadning.
3. This was also due to my misunderstanding. The gradient shape should be identical while it's summed and averaged across all ranks. - https://medium.com/@niruthiha2000/allreduce-explained-the-key-to-efficient-distributed-training-2cbbcc871832
4. Sorry for the inconvenience. It was mainly because our project is not public yet, containing a lot of ideas that shouldn't be public.

Again, I'm so sorry that you had to spend time because of my silly mistake🥲 and thank you for the interest in solving my issue!
